web/pandas/pdeps/0017-backwards-compatibility-and-deprecation-policy.md

```diff
@@ -0,0 +1,75 @@
+
+- Created: 27 June 2024
+- Status: Under discussion
+- Discussion: [#59125](https://github.com/pandas-dev/pandas/issues/59125)
+- Author: [Abdulaziz Aloqeely](https://github.com/Aloqeely)
+- Revision: 1
+
+## Abstract
+
+This PDEP defines pandas' backwards compatibility and deprecation policy.
+
+## Motivation
+
+Having a clear backwards compatibility and deprecation policy is crucial to having a healthy ecosystem. We want to ensure users can rely on pandas being stable while still allowing the library to evolve.
+
+This policy will ensure that users have enough time to deal with deprecations while also minimizing disruptions on downstream packages' users.
+
+## Scope
+
+This PDEP covers pandas' approach to backwards compatibility and the deprecation and removal process.
+
+## Background
+
+pandas uses a loose variant of semantic versioning.  
+A pandas release number is written in the format of ``MAJOR.MINOR.PATCH``.
+
+## General policy
+
+This policy applies to the [public API][1]. Anything not part of the [public API][1] or is marked as "Experimental" may be changed or removed at anytime.
+
+- Breaking backwards compatibility should benefit more than it harms users.
+- Breaking changes should go through a deprecation cycle before being implemented.
+- Breaking changes should only occur in major releases.
+- No deprecations should be introduced in patch releases.
+- Deprecated functionality should remain unchanged for at least 6 months before being changed or removed.
+
+Some bug fixes may require breaking backwards compatibility. In these cases, a deprecation cycle is not necessary. However, bug fixes which have a large impact on users might be treated as a breaking change. Whether or not a change is a bug fix or an API breaking change is a judgement call.
+
+## Deprecation process
+
+Deprecation provides a way to warn developers and give them time to adapt their code to the new functionality before the old behavior is eventually removed.
+
+A deprecation's warning message should:
+- Provide information on what is changing.
+- Mention how to achieve similar behavior if an alternative is available.
+- Include the version in which the deprecation will be enforced.
+- For large-scale deprecations, it is recommended to include a reason for the deprecation, alongside a discussion link to get user feedback.
+
+Additionally, when one introduces a deprecation, they should:
+- Use the appropriate warning class. More info on this can be found below.
+- Add the GitHub issue/PR number as a comment above the warning line.
+- Add an entry in the release notes.
+- Mention that the functionality is deprecated in the documentation using the ``.. deprecated::`` directive.
+
+### Which warning class to use
+
+Deprecations should initially use ``DeprecationWarning``, and then be switched to ``FutureWarning`` for broader visibility in the last minor release before the major release they are planned to be removed in.  
```

Comment:
Can we add "Significant" to the start of this? As it says, this PDEP says they should start as DeprecationWarning but don't have to, but don't provide any guidance on when they should/shouldn't.

Reply:
Agreed we could also use it, but I'd be quite strongly opposed. DeprecationWarning doesn't do what it seems most people think it does (users often still see it in their own code), and it would adds more maintenance on us.

Reply:
My understanding from the discussion at #54970 is that we should be starting with DeprecationWarning for most of our deprecations. 

> and it would add more maintenance on us.

There is not much added maintenance actually, all warnings should be using PandasDeprecationWarning which will be aliased to DeprecationWarning or FutureWarning depending on the version, you just have to modify the variable. (The test cases will be using the variable too)

Reply:
> DeprecationWarning doesn't do what it seems most people think it does (users often still see it in their own code)

Users see it in their code if they use a deprecated API directly (in which case, it seems right that they see it!), they just don't see it if a library they're using is using it (in which case, it's probably right that they don't see it, else as was pointed out in the linked discussion, dataframe-consuming libraries' users see warnings that they can't do anything about)

Given that the `PandasDeprecationWarning` trick seems to be low-maintenance, I think this is a net positive, and so am in favour

Reply:
Thanks for reminding me of that implementation @Aloqeely - agreed it's not as much maintenance. I'm positive here.